website/upgrade-guides/0-11.html.markdown

```diff
@@ -158,7 +158,109 @@ to explicitly pass those aliased providers.
 
 **Action**: Consider refactoring existing configurations so that all provider
 configurations are set in the root module and passed explicitly to child
-modules.
+modules, as described in the following section.
+
+### Moving Provider Configurations to the Root Module
+
+With the new provider inheritance model, it is strongly recommended to refactor
+any configuration where child modules define their own `provider` blocks so
+that all explicit configuration is defined in the _root_ module. This approach
+will ensure that removing a module from the configuration will not cause
+any provider configurations to be removed along with it, and thus ensure that
+all of the module's resources can be successfully refreshed and destroyed.
+
+A common configuration is where two child modules have different configurations
+for the same provider, like this:
+
+```
+# root.tf
+
+module "network-use1" {
+  source = "./network"
+  region = "us-east-1"
+}
+
+module "network-usw2" {
+  source = "./network"
+  region = "us-west-2"
+}
+```
+
+```
+# network/network.tf
+
+variable "region" {
+}
+
+provider "aws" {
+  region = "${var.region}"
+}
+
+resource "aws_vpc" "example" {
+  # ...
+}
+```
+
+The above example is problematic because removing either `module.network-use1`
+or `module.network-usw2` from the root module will make the corresponding
+provider configuration no longer available, as described in
+[issue #15762](https://github.com/hashicorp/terraform/issues/15762), which
+prevents Terraform from refreshing or destroying that module's `aws_vpc.example`
+resource.
+
+This can be addressed by moving the `provider` blocks into the root module
+as _additional configurations_, and then passing them down to the child
+modules as _default configurations_ via the explicit `providers` map:
+
+```
+# root.tf
+
+provider "aws" {
+  region = "us-east-1"
+  alias  = "use1"
+}
+
+provider "aws" {
+  region = "us-west-2"
+  alias  = "usw2"
+}
+
+module "network-use1" {
+  source = "./network"
+
+  providers = {
+    "aws" = "aws.use1"
+  }
+}
+
+module "network-usw2" {
+  source = "./network"
+
+  providers = {
+    "aws" = "aws.usw2"
+  }
+}
+```
+
+```
+# network/network.tf
+
+# No explicit "provider" blocks required here, because a default (non-aliased)
```

Comment:
Currently modules always requires an unconfigured provider block as a placeholder for a provider to be passed in, even if it doesn't have an alias.

Reply:
Oh, interesting... I'll play with this some more today and revise the docs. I'm a little worried that this change may be a bit _too_ breaking for existing modules (most modules in the registry won't have explicit provider blocks) so we may need to make an exception here for now until we get a chance to update existing modules, but I'll give it a try and see how it works out.

Reply:
I don't think it's a breaking change per say, since this functionality didn't exist for modules previously. Not having a provider config still works for implicit providers, and will use a matching top-level provider.

While it's possible to do, changing that won't be a quick fix. Missing providers can't be added until the configuration is resolved, but the provider inheritance is being setup from the configuration.